itest/rust/src/dictionary_test.rs

```diff
@@ -400,3 +372,279 @@ fn dictionary_equal() {
     assert_eq!(dict! {1: f32::NAN}, dict! {1: f32::NAN}); // yes apparently godot considers these equal
     assert_ne!(dict! {"foo": "bar"}, dict! {"bar": "foo"});
 }
+
+#[itest]
+fn dictionary_iter_create() {
+    let dictionary = dict! {
+        "foo": 0,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+
+    dictionary.iter_shared();
+    dictionary.keys_shared();
+}
+
+#[itest]
+fn dictionary_iter() {
+    let dictionary = dict! {
+        "foo": 0,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+
+    let map = HashMap::<String, Variant>::from([
+        ("foo".into(), 0.to_variant()),
+        ("bar".into(), true.to_variant()),
+        ("baz".into(), "foobar".to_variant()),
+        ("nil".into(), Variant::nil()),
+    ]);
+    for (key, value) in dictionary.iter_shared().typed::<String, Variant>() {
+        assert!(map.contains_key(&key), "key = \"{key}\"");
+        assert_eq!(map[&key], value, "key = \"{key}\"");
+    }
+
+    let map2: HashMap<_, _> = dictionary
+        .iter_shared()
+        .typed::<String, Variant>()
+        .collect();
+    assert_eq!(map, map2);
+    assert!(map2.contains_key("foo"));
+    assert!(map2.contains_key("bar"));
+    assert!(map2.contains_key("baz"));
+    assert!(map2.contains_key("nil"));
+}
+
+#[itest]
+fn dictionary_iter_equals_big() {
+    let dictionary: Dictionary = (0..1000).zip(0..1000).collect();
+    let map: HashMap<i64, i64> = (0..1000).zip(0..1000).collect();
+    let collected_map: HashMap<i64, i64> = dictionary.iter_shared().typed::<i64, i64>().collect();
+    assert_eq!(map, collected_map);
+    let collected_dictionary: Dictionary = collected_map.into_iter().collect();
+    assert_eq!(dictionary, collected_dictionary);
+}
+
+// Insertion mid-iteration is supported by godot
+
+#[itest]
+fn dictionary_iter_insert() {
+    let dictionary = dict! {
+        "foo": 0,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+    let mut dictionary2 = dictionary.share();
+
+    let mut iter = dictionary.iter_shared();
+    iter.next();
+    iter.next();
+
+    dictionary2.insert("new_key", 10);
+    let v: Vec<_> = iter.collect();
+    assert_eq!(dictionary.len(), 5);
+    assert!(dictionary.contains_key("new_key"));
+    assert_eq!(v.len(), 3);
+    assert!(v.contains(&("new_key".to_variant(), 10.to_variant())));
```

Comment:
These are great checks, thanks a lot! 🙂 

With them, we have a good chance of recognizing when the Godot implementation changes and our assumptions are no longer valid.